nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,439 @@
+    // VMM becomes the active VMM.
+    MIGRATION_UPDATE_INSTANCE -> "migration_update_instance" {
+        + siu_migration_update_instance
+    }
+
+    // Update network configuration to point to the new active VMM.
+    MIGRATION_UPDATE_NETWORK_CONFIG -> "migration_update_network_config" {
+        + siu_migration_update_network_config
+    }
+
+    // === active VMM destroyed actions ===
+
+    // Deallocate physical sled resources reserved for the destroyed VMM, as it
+    // is no longer using them.
+    DESTROYED_RELEASE_SLED_RESOURCES -> "destroyed_vmm_release_sled_resources" {
+        + siu_destroyed_release_sled_resources
+    }
+
+    // Deallocate virtual provisioning resources reserved by the instance, as it
+    // is no longer running.
+    DESTROYED_RELEASE_VIRTUAL_PROVISIONING -> "destroyed_vmm_release_virtual_provisioning" {
+        + siu_destroyed_release_virtual_provisioning
+    }
+
+    // Unassign the instance's Oximeter producer.
+    DESTROYED_UNASSIGN_OXIMETER_PRODUCER -> "destroyed_vmm_unassign_oximeter" {
+        + siu_destroyed_unassign_oximeter_producer
+    }
+
+    // Notify the V2P manager background task to delete the destroyed VMM's V2P
+    // mappings, and delete the destroyed VMM's NAT entries.
+    DESTROYED_UPDATE_NETWORK_CONFIG -> "destroyed_update_network_config" {
+        + siu_destroyed_update_network_config
+    }
+
+    DESTROYED_UPDATE_INSTANCE -> "destroyed_vmm_update_instance" {
+        + siu_destroyed_update_instance
+    }
+
+    DESTROYED_MARK_VMM_DELETED -> "destroyed_mark_vmm_deleted" {
+        + siu_destroyed_mark_vmm_deleted
+    }
+}
+
+// instance update saga: definition
+struct SagaDoActualInstanceUpdate;
+
+impl NexusSaga for SagaDoActualInstanceUpdate {
+    const NAME: &'static str = "instance-update";
+    type Params = RealParams;
+
+    fn register_actions(registry: &mut ActionRegistry) {
+        instance_update_register_actions(registry);
+    }
+
+    fn make_saga_dag(
+        params: &Self::Params,
+        mut builder: DagBuilder,
+    ) -> Result<steno::Dag, super::SagaInitError> {
+        fn const_node(
+            name: &'static str,
+            value: &impl serde::Serialize,
+        ) -> Result<steno::Node, super::SagaInitError> {
+            let value = serde_json::to_value(value).map_err(|e| {
+                SagaInitError::SerializeError(name.to_string(), e)
+            })?;
+            Ok(Node::constant(name, value))
+        }
+
+        builder.append(Node::action(
+            INSTANCE_LOCK_ID,
+            "GenerateInstanceLockId",
+            ACTION_GENERATE_ID.as_ref(),
+        ));
+        builder.append(become_updater_action());
+
+        // TODO(eliza): perhaps the start saga could determine whether it even
+        // needs to create a "real" saga based on whether either of the relevant
+        // state changes have occurred?
+
+        // If the active VMM's state has changed to "destroyed", then clean up
```

Comment:
I might consider structuring saga construction a bit differently. Given an `InstanceSnapshot` and all the information therein, we ought to be able to figure out:

- the posterior Instance record we want to write back to the DB
- the set of other actions we need to take to reconcile the rest of the system with the state that's about to be written back, namely:
  - networking updates:
    - if the active Propolis ID changed to another ID, the saga needs to write new networking config reflecting that ID
    - likewise, if it changed to None, we need to destroy existing config
    - also, if a migration is being retired, we need to poke the V2P background task, even if the migration failed
  - if any Propolis is now Destroyed or Failed, the saga needs to release its resources, destroy its Oximeter producers, etc.

 I would almost be inclined to have an associated function on `InstanceSnapshot` that produces as output the set of things that need to happen:

- do such-and-such networking update
- clean up resources for these VMMs
- write back this instance record

and then use that list to determine what saga is ultimately constructed. I think this might help avoid issues like the one I mention below w.r.t. deleting and then recreating instance networking state. But it might also just be a distinction without a difference, i.e. just another way of representing the same general decisions we're already making here. WDYT?

Reply:
Yeah, I think this seems like a good idea, I'll give it some more thought.

Reply:
> * if any Propolis is now Destroyed or Failed, the saga needs to release its resources, destroy its Oximeter producers, etc.

If I recall correctly, our original plan was _not_ to clean up `Destroyed`/`Failed` target VMMs in the update saga, and instead just unlink them and fail the _migration_, leaving the target to be cleaned up by the abandoned VMM reaper instead. The idea was that we would like to be able to start a new migration as soon as possible if a migration target died, and we didn't want to block that on completely cleaning up the old target[^1]. Instead, we decided that the saga is only responsible for cleaning up after _active_ VMM failures/destructions, because that actually has to happen within the instance lock. Is that still the intent?

[^1]: Which, in an eventual future world where we are able to mark VMMs as "failed" if their sleds fall over, may also no longer be possible! The target VMM may have failed because the whole sled it was on is gone, so it may be impossible to release some of its resources?

Reply:
I hadn't thought about it too hard when I dashed off that comment, but here's my take, per our offline discussion:

For non-migrating instances, we definitely want Nexus to clean up all the instance's resource counters (active VMM sled resources + virtual provisioning counters) before the instance goes back to the `NoVmm` state, in part to ensure that stopped instances don't ever appear to be charged against a quota, and in part to ensure an instance's virtual provisioning counters[^1] are gone before the next instance start saga runs.

Migration targets are a little murkier because it's less obvious (to external users) how having a migrating instance affects the system: there are no additional provisioning counters taken when an instance migrates, so users don't get charged more quota, and the utilization graphs in the console don't change. But (as we discussed offline) I think it's still important to clean up a migration's marginal resource usage before a new migration is allowed to start: otherwise, if there's an instance that's pathologically unable to migrate, and someone can trigger new migrations faster than the (failed) target VMMs can be cleaned up, it's possible to create a large number of defunct VMM resource reservations, preventing new instances from starting. I think the value of avoiding this outweighs the value of being able to start a new migration a bit more quickly after a failure (especially because, if one migration fails and the next succeeds, the total time spent getting the instance moved will probably be dominated by the time spent successfully migrating).

So I'd probably go for cleaning up any defunct VMMs we see in this path (which may be different from what I had originally thought!--I'd have to back and check my notes from March).

[^1]: I want to call them VPCs but CAN'T because that name is TAKEN gosh dang it

Reply:
Yup, thanks for writing up the conclusions from our conversation about this. I'll update the saga to always clean up any VMMs that have been destroyed when it snapshots the instance's state.

Reply:
This was a good idea, and I think we have a nice implementation of it now!